### ISSUE TITLE
TTS Engine Functionality not working

### ISSUE THREAD

**Issue by CavalierKingCharles** [GrapheneOS/os_issue_tracker#377]
From F-Droid, RHVoice TTS engine was installed. Voices could be installed in this. However, TTS functionality does not work in OsmAnd application. Sound was turned on of course! In the "Accessibility" settings under "Screen Readers" I could see that "RHVoices" was installed as the TTS engine. However, the pitch and speed and the word "PLAY" was greyed out. Uninstalling and re-installing RHVoices made things worse! Selecting TTS engine under "Screen Readers" as before, just causes the screen to keep reverting to the base settings menus; you can't even see the screen you saw before. You can't even see that RHVoices is installed as the TTS engine. From the IRC chat, I see that others have tried other TTS engines, but with no luck either.

**Comment by thestinger** [comment]
Is it targeting Android 11 without properly updating for it?

https://developer.android.com/about/versions/11/behavior-changes-11#tts-engines

**Comment by thestinger** [comment]
Please test if these work on the stock OS with Android 11.

**Comment by thestinger** [comment]
Please test if these work on the stock OS with Android 11.

If not, report the bugs to the app developers rather than here.

**Comment by thestinger** [comment]
There have been various upstream fixes and some compatibility improvements for the exec spawning feature since this was filed. Need to know fi you're still having any trouble with this.

**Comment by CavalierKingCharles** [comment]
Daniel,

Have just given it a shot. Unfortunately, though I have RHVoice installed from F-droid and the voice working, and getting the text notifications appearing, the TTS functionality still is not working.

Just to say, appreciate all work you’ve done on GrapheneOS.  Have been using as DD for some time. Totally Apple and Google free now. More power to your elbow.

**Comment by thestinger** [comment]
Please test on the 2020.12.08.08 release.

**Comment by CavalierKingCharles** [comment]
Have just tested on the latest release just deployed to my phone - the one indicated above. Sadly, there are still no voice directions in OsmAnd. I'm still seeing the text directions in the notifications.  Voice confirmed as downloaded and working in RHVoice.

Please would someone confirm all the settings in the phone to enable this to work. I'm pretty sure I have done all I can, but maybe there's something I've missed? Is this functionality working on other Pixels rather than my 3? Or is it universal?

**Comment by thestinger** [comment]
Have you selected it as the TTS engine in the OS settings?

**Comment by CavalierKingCharles** [comment]
Mmm - in "Accessibility", there is a section entitled "SCREEN READERS", in which we see "Text-to-speech output". When I select this, the screen goes blank for a moment and returns to the main Settings screen. Is there anywhere else I can select RHVoice as the TTS engine?

**Comment by thestinger** [comment]
Get the output from `adb logcat` for when you trigger that. Also, have you tried using any other text-to-speech apps? This is likely a problem with this app.

**Comment by CavalierKingCharles** [comment]
Will do...

**Comment by TriggerDingus** [comment]
Just a note that I am encountering the precise issues noted by the OP. I'm on Pixel 4, build ...12.12.03.
I encounter the same behavior as in the original post with RH Voice. I tried other TTS engines (such as Flite), but they seem outdated and/or crash altogether.

**Comment by thestinger** [comment]
No logs have been provided yet.

**Comment by TriggerDingus** [comment]
My programming days are more than 15 years behind me so gathering logs was a tricky escapade for me, therefore I can't guarantee the result. Please find attached.
[logcat.txt](https://github.com/GrapheneOS/os_issue_tracker/files/5697934/logcat.txt)

**Comment by CavalierKingCharles** [comment]
[logcat.txt](https://github.com/GrapheneOS/os_issue_tracker/files/5703978/logcat.txt)

Daniel, Sorry for delay. As Trigger Dingus, I tried a number of TTS apps. Same result for all of them. This logfile should contain where I mentioned "Selecting TTS engine under "Screen Readers" as before, just causes the screen to keep reverting to the base settings menus; you can't even see the screen you saw before"  It should also contain where I started OsmAnd and it said "unable to start TTS engine"

**Comment by thestinger** [comment]
We've confirmed that Flite TTS, RHVoice and Google TTS can be successfully set as the active TTS engine on GrapheneOS and work inside the Settings app. You need to install the relevant voices before RHVoice will work. That was the original issue that you had and the issue is simply that you didn't install the voices. TTS engine functionality is working correctly. This issue can instead be about the issue you've run into where you uninstalled the configured TTS engine and now the menu isn't opening.

**Comment by thestinger** [comment]
OSMand has also been confirmed to work properly when TTS has been properly configured.

**Comment by thestinger** [comment]
The correct place to configure TTS is Settings > Accessibility > Text to speech output. The option won't appear if there is no available TTS engine.

**Comment by CavalierKingCharles** [comment]
I’m sorry Daniel, but I did install the voices many times for each TTS engine I installed.  I’ll give this one more shot.

**Comment by CavalierKingCharles** [comment]
I have done that more times than I have had hot dinners this last month!

**Comment by thestinger** [comment]
Has also been confirmed that it can be used successfully in various apps include OSMand and an ebook reader. It appears that the issue here is something to do with setting it up in a way that breaks Settings.

**Comment by CavalierKingCharles** [comment]
Just tried again on RHVoice. Voices downloaded. I get the audible message “if you can hear this message, the voice is installed and working”

As I reported when I first raised this issue, I go to Settings - Accessibility- Text to speech output, but it falls over and returns to the Settings menu.

Nothing has changed, I have installed, re-installed various TTS engines over and over again. Voices installed as well. My Pixel 3 still gets the message “cannot start TTS engine” when OsmAnd starts and still falls over when in settings menu as described above.

Did you look at my logs that I sent you when I repeated these operations with these results? What did they say?

**Comment by CavalierKingCharles** [comment]
What steps in what order are necessary to get this working?  I’m not sure I can see another way!

**Comment by thestinger** [comment]
@CavalierKingCharles Are you on blueline (non-XL Pixel 3)? I can make a test release with a potential workaround for the issue for you.

**Comment by thestinger** [comment]
@CavalierKingCharles Are you on blueline (non-XL Pixel 3)? I can make an Alpha release with a potential workaround for the issue for you.

**Comment by thestinger** [comment]
@CavalierKingCharles

Here's the Alpha release to test:

https://releases.grapheneos.org/blueline-ota_update-2020.12.21.02.zip

This may work around this problem or at least allow the code to progress further and throw an error at a later point. Let me know how it goes.

Changes since the 2020.12.12.03 release:

* Settings: update GrapheneOS connectivity check URLs to match NetworkStack
* Camera: remove unused Wi-Fi state permissions
* Pixel 3, Pixel 3 XL, Pixel 3a, Pixel 3a XL, Pixel 4, Pixel 4 XL, Pixel 4a: update APNs with carriersettings-extractor
* adjust kernel configuration tests to permit not having BPF_JIT since we don't have it enabled
* add check for empty TTS engine name to address upstream bug

**Comment by CavalierKingCharles** [comment]
Yup, non-XL Pixel 3

On 21 December 2020 02:33:50 GMT, Daniel Micay <notifications@github.com> wrote:
>@CavalierKingCharles Are you on blueline (non-XL Pixel 3)? I can make
>an Alpha release with a potential workaround for the issue for you.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/GrapheneOS/os_issue_tracker/issues/377#issuecomment-748723223

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by thestinger** [comment]
Can you try sideloading this Alpha OTA update?

https://grapheneos.org/usage#updates-sideloading

**Comment by CavalierKingCharles** [comment]
I’ll give it a shot Daniel.  It may not be in the next few days though due to commitments. I may need to wait until “dead time” after Christmas Day.

**Comment by CavalierKingCharles** [comment]
Have sideloaded Alpha OTA update as suggested. Good news - we DO get further. In: Settings  > Accessibility > Text to speech output we are now able to select "Text-to-speech output (before we just blew out back to Settings menu).  Under "Preferred Engine", if I tap that, I see "RHVoice" selected with a radio button.
Unfortunately, when starting up OsmAnd, I still get the message "Cannot Start text-to-speech engine.
Progress though as you had hoped.
I await further instructions.

**Comment by thestinger** [comment]
Does RHVoice work within the Settings app?

**Comment by CavalierKingCharles** [comment]
"Speech rate" and "Pitch" are greyed out, as is "Play".  Only "Reset" is green.  I therefore presume the answer is "No".

**Comment by thestinger** [comment]
Can you try figuring that out by uninstalling it, reinstalling it, installing voices again, etc.

Also, consider trying another one like flite at least for debugging this.

**Comment by CavalierKingCharles** [comment]
will do...over the pond as you have guessed, so much later here. I'll let you know later on if don't fall asleep or first thing tomorrow.

**Comment by CavalierKingCharles** [comment]
Daniel - Uninstalled and re-installed RHVoice. Installed voice. Went to setting and selected RHVoice again. When selected back arrow, we now have a cog against "RHVoice", neither of which was there before. Am able to update speech rate etc.
Went outside - NO message "Cannot start text-to-speech engine".  Selected a destination in OsmAnd and started the route.
Voice directions now functioning!  An early Christmas present!! Not that I can go anywhere in miserable locked down UK, but ready to roll!! Great stuff!

**Comment by thestinger** [comment]
Okay, I think the fix that I applied for the upstream issue resolves this problem. The issue seems to be that they didn't handle the case of not having any TTS providers properly.

**Comment by sonora** [comment]
@thestinger I authored a good chunk of the OsmAnd voice prompting code, and just discovered this.

Would you have any idea what in the code could be missing regarding '_didn't handle the case of not having any TTS providers properly_'? It seems we have a user wirh a rather special systen setup having a similar issue (https://github.com/osmandapp/OsmAnd/issues/9637).

**Comment by thestinger** [comment]
@sonora AOSP doesn't come with a TTS provider and therefore none available until the user installs and enables one. This can also occur even on an OS with Play services included. You could test with AOSP emulator images not containing Play services where there won't be a TTS provider included by default. The app should handle absence of an enabled OS TTS implementation gracefully without crashing or producing a strange error. It could recommend that users install RHVoice or espeak-ng on a system without Play services.

The confusion here seemed to be that when you install a TTS engine, it shows up as selected in the settings since it's the only one, but you still need to press on it there to activate it. The OS Settings UI for this was not designed to make it clear when no TTS engine is enabled. It just assumes that if there's 1 TTS engine, that's what is enabled. Pressing it in that menu works fine and allows activating it.

**Comment by sonora** [comment]
Thank you so much for he reply! In the issue I reference above the user seems to in deed talk about a system where all native TTS services had been removed, and he now says for all post-installed TTS engines he receives "Could not start TTS engine". So it may well be an issue of some activation missing in the system settings.I am currently not sure how to positively identify such situations in app code in order to provide better user guidance. But thx again, anyway!

**Comment by thestinger** [comment]
The user needs to explicitly activate the TTS engine they want to use in the OS settings. They might be opening up the OS settings and looking at it, seeing that it's selected, and thinking that it's activated. It isn't until they press it there and accept the dialog for activating it.

**Comment by sonora** [comment]
@thestinger Many thanks!
@seniorm0ment Maybe this helps it?

**Comment by awebeer256** [comment]
> The confusion here seemed to be that when you install a TTS engine, it shows up as selected in the settings since it's the only one, but you still need to press on it there to activate it. The OS Settings UI for this was not designed to make it clear when no TTS engine is enabled. It just assumes that if there's 1 TTS engine, that's what is enabled. Pressing it in that menu works fine and allows activating it.

A year and a half later, this problem in the Settings app still hasn't been fixed. I installed GrapheneOS and RHVoice, but then the Settings app's top activity would crash whenever I tried to select a voice language. Sure enough, I had to select RHVoice as the engine to use first, even though it looked like it was already selected!

Has this issue been submitted to upstream (AOSP)? If so, can we get a link?

**Comment by thestinger** [comment]
You're welcome to file an upstream issue.
